File: app/lib/models/spotless_parameters.dart
import 'package:json_annotation/json_annotation.dart';

part 'spotless_parameters.g.dart';

/// Which spot remover to run.
enum SpotLessMethod {
  @JsonValue('spotless')
  spotless('SpotLess'),
  @JsonValue('removeDirt')
  removeDirt('RemoveDirt (fast)');

  const SpotLessMethod(this.displayName);
  final String displayName;
}

/// Parameters for the SpotLess pass.
/// Removes dust, dirt, and temporal spots using motion-compensated median.
@JsonSerializable()
class SpotLessParameters {
  /// Whether this pass is enabled.
  final bool enabled;

  /// Which spot remover to run.
  final SpotLessMethod method;

  /// RemoveDirt tuning.
  final int rdGmthreshold;
  final int rdNoise;
  final int rdNoisy;
  final int rdDist;
  final bool rdPostDenoise;

  /// Process chroma planes (default true).
  final bool chroma;

  /// Recalculate motion vectors for more precision (default false).
  final bool rec;

  /// Block size for motion analysis (default 16).
  final int blksize;

  /// Block overlap (default 8).
  final int overlap;

  /// Sub-pixel accuracy: 1=pixel, 2=half, 4=quarter (default 2).
  final int pel;

  const SpotLessParameters({
    this.enabled = false,
    this.method = SpotLessMethod.spotless,
    this.rdGmthreshold = 70,
    this.rdNoise = 50,
    this.rdNoisy = 12,
    this.rdDist = 1,
    this.rdPostDenoise = false,
    this.chroma = true,
    this.rec = false,
    this.blksize = 16,
    this.overlap = 8,
    this.pel = 2,
  });

  SpotLessParameters copyWith({
    bool? enabled,
    bool? chroma,
    bool? rec,
    int? blksize,
    int? overlap,
    int? pel,
    SpotLessMethod? method,
    int? rdNoise,
    int? rdNoisy,
    int? rdGmthreshold,
    int? rdDist,
    bool? rdPostDenoise,
  }) {
    return SpotLessParameters(
      enabled: enabled ?? this.enabled,
      chroma: chroma ?? this.chroma,
      rec: rec ?? this.rec,
      blksize: blksize ?? this.blksize,
      overlap: overlap ?? this.overlap,
      pel: pel ?? this.pel,
      method: method ?? this.method,
      rdNoise: rdNoise ?? this.rdNoise,
      rdNoisy: rdNoisy ?? this.rdNoisy,
      rdGmthreshold: rdGmthreshold ?? this.rdGmthreshold,
      rdDist: rdDist ?? this.rdDist,
      rdPostDenoise: rdPostDenoise ?? this.rdPostDenoise,
    );
  }

  /// Get a summary string for display.
  String get summary {
    if (!enabled) return 'Off';
    final parts = <String>[];
    if (rec) parts.add('Refined');
    if (!chroma) parts.add('Luma only');
    return parts.isEmpty ? 'SpotLess' : 'SpotLess (${parts.join(", ")})';
  }

  factory SpotLessParameters.fromJson(Map<String, dynamic> json) =>
      _$SpotLessParametersFromJson(json);
  Map<String, dynamic> toJson() => _$SpotLessParametersToJson(this);
}

File: worker/src/models/spotless_parameters.rs
use serde::{Deserialize, Serialize};

/// Which spot remover to run.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
pub enum SpotLessMethod {
    /// SpotLess — motion-compensated temporal median. Higher quality, slow.
    #[default]
    #[serde(rename = "spotless")]
    SpotLess,
    /// RemoveDirt — measured 6.3x faster for about 60% of the removal.
    #[serde(rename = "removeDirt")]
    RemoveDirt,
}

/// Parameters for the SpotLess pass.
/// Removes dust, dirt, and temporal spots using motion-compensated median.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpotLessParameters {
    /// Whether this pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Which spot remover to run.
    #[serde(default)]
    pub method: SpotLessMethod,

    /// RemoveDirt: global motion threshold — how much of the frame may differ
    /// before it is treated as motion rather than damage.
    #[serde(default = "default_rd_gmthreshold")]
    pub rd_gmthreshold: i32,
    /// RemoveDirt: how large a difference counts as a spot.
    #[serde(default = "default_rd_noise")]
    pub rd_noise: i32,
    /// RemoveDirt: how many neighbouring pixels must agree.
    #[serde(default = "default_rd_noisy")]
    pub rd_noisy: i32,
    /// RemoveDirt: dilation distance around a detected spot.
    #[serde(default = "default_rd_dist")]
    pub rd_dist: i32,
    /// RemoveDirt: re-enable the canonical trailing RemoveGrain(17). Off by
    /// default — measured, it alone triples the collateral damage.
    #[serde(default)]
    pub rd_post_denoise: bool,

    /// Process chroma planes (default true).
    #[serde(default = "default_true")]
    pub chroma: bool,

    /// Recalculate motion vectors for more precision (default false).
    #[serde(default)]
    pub rec: bool,

    /// Block size for motion analysis (default 16).
    #[serde(default = "default_blksize")]
    pub blksize: i32,

    /// Block overlap (default 8).
    #[serde(default = "default_overlap")]
    pub overlap: i32,

    /// Sub-pixel accuracy: 1=pixel, 2=half, 4=quarter (default 2).
    #[serde(default = "default_pel")]
    pub pel: i32,
}

fn default_true() -> bool { true }
fn default_blksize() -> i32 { 16 }
fn default_overlap() -> i32 { 8 }
fn default_pel() -> i32 { 2 }

fn default_rd_gmthreshold() -> i32 { 70 }
fn default_rd_noise() -> i32 { 50 }
fn default_rd_noisy() -> i32 { 12 }
fn default_rd_dist() -> i32 { 1 }

impl Default for SpotLessParameters {
    fn default() -> Self {
        Self {
            enabled: false,
            chroma: default_true(),
            rec: false,
            blksize: default_blksize(),
            overlap: default_overlap(),
            pel: default_pel(),
            method: SpotLessMethod::default(),
            rd_gmthreshold: default_rd_gmthreshold(),
            rd_noise: default_rd_noise(),
            rd_noisy: default_rd_noisy(),
            rd_dist: default_rd_dist(),
            rd_post_denoise: false,
        }
    }
}
